swarm/metrics.py

```diff
@@ -0,0 +1,165 @@
+__author__ = "Aidan Morrison <aidandmorrison@gmail.com>"
+
+import numpy as np
+
+
+def summarise_across(data:np.ndarray, func_string: str = 'mean', across: str = "bees") -> np.ndarray:
```

Comment:
So `np.mean(data, axis=0) == summarise_across(data, 'mean', 'bees')`? 

I think you're on the right path - this would make code more readable, but I don't think it justifies it's complexity. If I wanted a function you didn't support, I'd be out of luck or I wanted to summarise across an axis that you didn't name, I wouldn't gain anything. This code spends almost it's entire time doing error checking, just to make a single `np.apply_along_axis` call.

I can see you're influenced by the `dplyr::summarise` function here, but I feel like it's a subpar fit - `dplyr::summarise` is fundamentally for data frames, not arrays/tensors so it requires less complexity since it is always summarising individual rows (hence why you need an extra arg).

Effectively what you'd like is something akin to http://xarray.pydata.org/en/stable/index.html where each axis is named. So you could do something like `np.mean(data, axis='bees')` which I also really like the idea of and maybe that might be a better approach to what you're doing here (converting results into named arrays for easier analysis?). 

I'm not overly in favour of this function as it currently stands tbh. One you get rid of the `range` logic, you're really only doing `bees = 0` and `x=-1` here.

Reply:
Actually, with a little sigh, I probably agree.  

I went into this thinking I'd just figure out what I wanted to do, and build as many functions as seemed apt to do it. This function started out as only summarising across bees, strictly to get the 'prediction of the swarm' from the swarm of predictions.  I actually almost went about this a far, far uglier way, but a little research revealed `np.apply_along_axis`, which fitted the job so perfectly that I did pause for a moment to reconsider whether it was even worth a function.  But since I had range on my list of functionality and didn't stumble across `np.ptp`, I kept it. 
When I realised that specifying different axes actually took care of the other use cases too, I thought maybe it being a function was justified.  But I agree, it's really marginal. And xarray looks pretty cool, and a much more 'general' yet fit for purpose solution.  

On the other hand, part of me questions whether calling a new library just for this use case is justified. On the other hand, it might be, and in other circumstances.  The problem with just calling `np.apply_along_axis` with two different axes indicated by an integer is that the code is so concise that it verges on obscuring the actual rationale for what it's doing.  I don't know the ideal solution for that.  Perhaps just better comments in code. But `np.mean(data, axis = 'bees')` looks pretty damn good too.